python_modules/dagster/dagster/_core/definitions/sensor_definition.py

```diff
@@ -154,6 +160,205 @@ def repository_name(self) -> Optional[str]:
+        """The cursor value for this sensor, which was set in an earlier sensor evaluation."""
+        return self._cursor
+
+    @public
+    def update_cursor(self, cursor: Optional[Mapping[AssetKey, Optional[int]]]) -> None:
+        """Updates the cursor value for this sensor, which will be provided on the context for the
+        next sensor evaluation.
+
+        This can be used to keep track of progress and avoid duplicate work across sensor
+        evaluations.
+
+        Args:
+            cursor (Optional[Mapping[AssetKey, int]]):
+        """
+        self._cursor = check.opt_dict_param(cursor, "cursor", key_type=AssetKey)
+
+    @public  # type: ignore
+    @property
+    def repository_name(self) -> Optional[str]:
+        return self._repository_name
+
+    @public
+    def latest_materialization_records_by_key(
+        self, asset_keys: Optional[Sequence[AssetKey]] = None
+    ) -> Mapping[AssetKey, Optional["EventLogRecord"]]:
+        """Fetches the most recent materialization event record for each asset in asset_keys.
+
+        Args:
+            asset_keys (Optional[Sequence[AssetKey]]): list of asset keys to fetch events for. If not specified, the
+                latest materialization will be fetched for all assets the multi_asset_sensor monitors.
+
+        Returns: Mapping of AssetKey to EventLogRecord where the EventLogRecord is the latest materialization event for the asset. If there
+            is no materialization event for the asset, the value in the mapping will be None.
+        """
+        from dagster._core.events import DagsterEventType
+        from dagster._core.storage.event_log.base import EventRecordsFilter
+
+        if asset_keys is None:
+            asset_keys = self._asset_keys
+
+        cursor_dict = self.cursor if self.cursor else {}
+        asset_event_records = {}
+        for a in asset_keys:
+            event_records = self.instance.get_event_records(
+                EventRecordsFilter(
+                    event_type=DagsterEventType.ASSET_MATERIALIZATION,
+                    asset_key=a,
+                    after_cursor=cursor_dict.get(a),
+                ),
+                ascending=False,
+                limit=1,
+            )
+
+            if event_records:
+                asset_event_records[a] = event_records[0]
+            else:
+                asset_event_records[a] = None
+
+        return asset_event_records
+
+    @public
+    def materialization_records_for_key(
+        self, asset_key: AssetKey, limit: int, ascending: bool = True
+    ) -> Iterable["EventLogRecord"]:
+        """Fetches asset materialization event records for asset_key
+
+        Args:
+            asset_key (AssetKey): The asset to fetch materialization events for
+            limit (int): The number of events to fetch
+            ascending (bool): If True, materialization events will be returned with the earliest event first. If False, the most recent
```

Comment:
I think the cursor logic might behave in a surprising way if we allow for `ascending=False`, because when we pass the event into `advance_cursor`, it will filter to events AFTER that cursor, which we've already examined.  I'm actually fine removing the `ascending` parameter so we don't have to deal with it, but it's worth thinking through if our cursor API has the right set of parameters also.

Reply:
oh i see what you mean. I think we'd need to expose the `before_cursor` arg on the `EventsFilter` too and keep track of two cursors per assetkey. I think that would be a case where it would be reasonable to tell the user to write a plain `sensor` that does the event fetching manually